GDScript: Report property type errors

ZuBsPaCe opened the pull request:
Before, inline getters and setters were of type SuiteNode which were not checked by the gdscript analyzer. resolve_class_body only checked FunctionNode members.

To fix #53102 I transformed those inline getters and setters into FunctionNodes. Please check the example code provided in the issue.

Additionally, I also added some missing checks to old-style getset-properties:
* Getter must return a value and have zero parameters. If the property has a type, the returned type of the getter must be a variant or a compatible type.
* Setter must have a single parameter. If the property has a type, the parameter must be a variant or a compatible type.
* If the property is untyped, either the getter return type or the setter parameter must be a variant or their type must be compatible.

ZuBsPaCe committed:
GDScript: Report property type errors

Inline getters & setters are now FunctionNodes.
Their names are set in the parser, not in the compiler.
GDScript-Analyzer will now run through getter and setter.
Also report wrong type or signature errors regarding getset properties.
Added GDScript tests for getters and setters.
#53102

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -869,18 +869,32 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class) {
+			for (GDScriptParser::AnnotationNode *&E : member.function->annotations) {
+				E->apply(parser, member.function);
+			}
+		} else if (member.type == GDScriptParser::ClassNode::Member::VARIABLE && member.variable->property != GDScriptParser::VariableNode::PropertyStyle::PROP_NONE) {
```
We usually don't add the enum name when referring to its elements. Although this is more of a legacy convention since older compilers didn't support it, given that in other places this code doesn't have the enum name (e.g. in `GDScriptParser::ClassNode::Member::FUNCTION` it doesn't specify the `Type` enum), I think it should be better to keep the current convention, so this should use `GDScriptParser::VariableNode::PROP_NONE`.

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -896,17 +910,80 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class) {
+			}
+#endif
+
+			if (member.type == GDScriptParser::ClassNode::Member::VARIABLE && member.variable->property == GDScriptParser::VariableNode::PropertyStyle::PROP_SETGET) {
```
The condition `member.type == GDScriptParser::ClassNode::Member::VARIABLE` is already checked in the outer `if`, so it's not needed here.

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1518,12 +1595,24 @@ void GDScriptAnalyzer::resolve_parameter(GDScriptParser::ParameterNode *p_parame
+		expected_type = parser->current_function->get_datatype();
+		has_expected_type = true;
+	}
+	//else if (parser->current_property != nullptr) {
+	//	expected_type = parser->current_property->get_datatype();
+	//	has_expected_type = true;
+	//}
```
Any reason to add commented out code?

vnen commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -947,17 +947,35 @@ GDScriptParser::VariableNode *GDScriptParser::parse_property(VariableNode *p_var
-			p_variable->setter = parse_suite("setter definition");
-			break;
+			FunctionNode *function = alloc_node<FunctionNode>();
+			function->identifier = p_variable->identifier;
```
I think this might be a bit dangerous considering that you temporarily change the name in the compiler. It's better to create a new `IdentifierNode` and set its name here.

vnen commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -947,17 +947,35 @@ GDScriptParser::VariableNode *GDScriptParser::parse_property(VariableNode *p_var
+			SuiteNode *body = alloc_node<SuiteNode>();
+			body->add_local(parameter, function);
+
+			function->body = parse_suite("function declaration", body);
```
The context should be `setter declaration` to make error messages clearer.

```suggestion
			function->body = parse_suite("setter declaration", body);
```

vnen commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -972,11 +990,22 @@ void GDScriptParser::parse_property_setter(VariableNode *p_variable) {
+			current_function = function;
+
+			SuiteNode *body = alloc_node<SuiteNode>();
+			function->body = parse_suite("function declaration", body);
```
```suggestion
			function->body = parse_suite("getter declaration", body);
```

vnen commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4433,7 +4462,7 @@ void GDScriptParser::TreePrinter::print_variable(VariableNode *p_variable) {
 				push_line(":");
 				increase_indent();
-				print_suite(p_variable->getter);
+				print_function(p_variable->getter);
```
This might be a bit awkward when printed out since it already prints a sort of signature for the getter, then `print_function` will add another signature. It's better to keep printing only the suite:

```suggestion
				print_suite(p_variable->getter->body);
```

Same thing for the setter.

vnen commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4453,7 +4482,7 @@ void GDScriptParser::TreePrinter::print_variable(VariableNode *p_variable) {
 				push_line("):");
 				increase_indent();
-				print_suite(p_variable->setter);
+				print_function(p_variable->setter);
```
```suggestion
				print_suite(p_variable->setter->body);
```

vnen reviewed (changes requested):
Overall looks good to me, I just have some nitpicks in the way it's done.

vnen commented:
One other thing: if you could add integration tests for those cases it would be great. See [the documentation](https://docs.godotengine.org/en/latest/development/cpp/unit_testing.html#integration-tests-for-gdscript) for info about how those work.

ZuBsPaCe commented:
@vnen Thanks for your detailed review! Things I did:

* Fixed the things in your comments exactly as you told me to. See Commit 649849c

* Added some test cases for properties. It's not a simple task, because there are so many things which could be tested. Adding tests for all different kinds of errors is cumbersome, because each test can only check one error. Added some simple tests in commit 1fdded2

* I rebased those 3 commits to master without squashing. Hope that's what you expected.

vnen reviewed (changes requested):
Other than the comments it's okay, could be merged after addressing those.

The commits can be squashed into one as well.
